syntax = "proto3";

package users;

option go_package = "./pb";

service UserService {
  rpc CreateUser(CreateUserRequest) returns (UserResponse);
  rpc AuthenticateUser(AuthRequest) returns (AuthResponse);
  rpc GetUser(GetUserRequest) returns (UserResponse);
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
  rpc UpdateUser(UpdateUserRequest) returns (UserResponse);
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
}

enum UserType {
  UNKNOWN_ROLE = 0;
  ADMINISTRADOR = 1;
  MEDICO = 2;
  RECEPCIONISTA = 3;
  PACIENTE = 4;
}

message CreateUserRequest {
  int32 token = 1;
  string name = 2;
  string email = 3;
  string password = 4;
  UserType user_type = 5;
}

message AuthRequest {
  string email = 1;
  string password = 2;
}

message AuthResponse {
  bool success = 1;
  int32 token = 2;
  int32 user_id = 3;
  UserType user_type = 4;
  string message = 5;
}

message GetUserRequest {
  int32 token = 1;
  int32 user_id = 2;
}

message ListUsersRequest {
  int32 token = 1;
  UserType user_type = 2; // Filtro por tipo de usuário.
  int32 limit = 3;
  int32 offset = 4;
}

message ListUsersResponse {
  repeated UserResponse users = 1;
  int32 total = 2;
}

message UpdateUserRequest {
  int32 token = 1;
  int32 user_id = 2; // ID de quem será alterado.
  string name = 3;
  string email = 4;
  string password = 5;
}

message UserResponse {
  int32 user_id = 1;
  string name = 2;
  string email = 3;
  UserType user_type = 4;
}

message DeleteUserRequest {
  int32 token = 1;
  int32 user_id = 2;
}

message DeleteUserResponse {
  bool success = 1;
  string message = 2;
}